### PR TITLE
Add more robust processing indicators (fixes #411)

### LINKED ISSUE #411: Show general payment processing errors
When processing a payment, the user does not see any general processing error, i.e. one not associated with a form field. They see either an empty credit card form or a spinner but no error message. This is pretty bad because there are a lot of errors that could happen like: https://developers.braintreepayments.com/javascript+python/reference/general/testing#unsuccessful-credit-card-verification

There is a UX mock for how this should look (the last screen here):
ux-pay-form-error

### PR DESCRIPTION
**Pull request by kumar303**
This adds processing state and also fixes broken forms after handling errors. 

Fixes: https://github.com/mozilla/payments-ui/issues/411

The approach:
- adds a flexible mechanism to store/retrieve processing state for any component
- any action should dispatch begin/stop notifications when processing data
  - I only implemented this for the credit card submission and saved pay method submission
- a lot of components are now watching state with `connect()` but for simplicity I mostly tested the raw class objects, not provider-wrapped objects. I think this works ok because dispatch isn't really involved in the tests.

Here is a screenshot of an enabled submit button after viewing errors which was not possible before. What isn't shown here is that the spinner only occurs during submission, then it goes away.

screenshot 2015-09-10 18 52 47

### CODE REVIEW

File: public/scss/_app-error.scss [hunk 1]
```diff
@@ -1,21 +0,0 @@
-.app-error {
```
Comment (kumar303) [inline public/scss/_app-error.scss]: I don't think this is used anymore, right?

File: public/js/utils.jsx [hunk 2]
```diff
@@ -169,6 +169,6 @@ export function arrayHasSubString(list, searchSubString) {
   return false;
 }
 
-export function getId() {
-  return '_' + Math.random().toString(36).substr(2, 9);
+export function getId({prefix=''} = {}) {
+  return prefix + '_' + Math.random().toString(36).substr(2, 9);
```
Comment (muffinresearch) [inline public/js/utils.jsx]: Nice!

File: public/scss/_app-error.scss (hunk 1 above)
Comment (muffinresearch) [inline public/scss/_app-error.scss]: Yep it's good to go.

**Comment by muffinresearch** [comment]
I suppose one thing about that specific example in the screenshot is that the user shouldn't be able to keep submitting as it will always fail. Maybe something to think about for the future is a different class of error notifications. E.g. showFatal which goes back to more of a modal error?

File: public/js/actions/processing.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+
+export function stopProcessing(processingId) {
+  return processingAction(processingId, actionTypes.STOP_PROCESSING);
+}
+
+
+function processingAction(processingId, actionType) {
+  if (!processingId) {
+    throw new Error('processingId cannot be empty');
```
Comment (muffinresearch) [inline public/js/actions/processing.js]: :+1:

File: public/js/components/card-form.jsx [hunk 4]
```diff
@@ -147,13 +150,12 @@ export default class CardForm extends Component {
...
     }
-    this.props.handleCardSubmit(formData, opts);
+    this.props.handleCardSubmit(formData, this.processingId, opts);
   }
 
   mapErrorsToFields(errors) {
     console.log('mapping submission errors to form fields', errors);
-    // TODO: map non-braintree errors like __all__
-    // TODO: map braintree errors for unexpected fields (such as `plan_id`)
+    // Note that __all__ errors will be rendered as notifications instead.
```
Comment (muffinresearch) [inline public/js/components/card-form.jsx]: Thx!

File: tests/views/test.product-pay.jsx [hunk 5]
```diff
@@ -10,21 +12,33 @@ describe('ProductPay', function() {
...
+              cardSubmissionErrors={{}}
+              productId={productId}
+              processPayment={fakePaymentProcessor}
+            />
+          );
+        }}
+      </Provider>
+    );
+
+    return TestUtils.findRenderedComponentWithType(
```
Comment (muffinresearch) [inline tests/views/test.product-pay.jsx]: Maybe we should have added a helper to do by Type just for consistency with all the other stuff we have.

File: public/js/views/management/add-pay-method.jsx [hunk 6]
```diff
@@ -33,7 +35,7 @@ export default class AddPayMethod extends Component {
         <CardForm
           submissionErrors={this.props.cardSubmissionErrors}
           submitPrompt={gettext('Add')}
-          handleCardSubmit={(card) => this.handleCardSubmit(card)}
+          handleCardSubmit={this.handleCardSubmit.bind(this)}
```
Comment (muffinresearch) [inline public/js/views/management/add-pay-method.jsx]: Could you make the handleCardSubmit a fat arrow func then you probably don't need the bind.

e.g:
```
handleCardSubmit = (creditCard, processingId) => {

}
```

File: tests/reducers/test.processing-reducer.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+                     initialProcessingState);
+  });
+
+  it('should store true when a component begins processing', function() {
+    var procId = 'some-id';
+    var state = processingReducer(initialProcessingState, {
+      type: actionTypes.BEGIN_PROCESSING,
+      processingId: procId,
+    });
+    assert.equal(state[procId], true);
```
Comment (muffinresearch) [inline tests/reducers/test.processing-reducer.js]: Could just used `assert.ok`.

**Comment by muffinresearch** [comment]
Excellent stuff! r+wc

File: tests/views/test.product-pay.jsx (hunk 5 above)
Comment (kumar303) [inline tests/views/test.product-pay.jsx]: I was thinking this Provider/find pattern could probably be wrapped up a little better too

File: public/js/views/management/add-pay-method.jsx (hunk 6 above)
Comment (kumar303) [inline public/js/views/management/add-pay-method.jsx]: one thing I like about bind is that you don't have to update it when the signature changes. In this case, it actually took me a while to track down that the lambda wasn't passing the right args :(

**Comment by kumar303** [comment]
> the user shouldn't be able to keep submitting as it will always fail

In the CardForm case, it has no way to know that the form is invalid which is why the button is enabled. One thing we could do is put a `wasSuccessful` property on the processing state. That would be a pretty simple fix. Maybe I'll try it here and punt to another patch if it gets complicated.

**Comment by kumar303** [comment]
> put a wasSuccessful property on the processing state

actually, nevermind. That's not the right place for it because processing state is ephemeral in nature.

**Pull request merged**